include/fmt/format.h

```diff
@@ -1822,85 +1843,201 @@ class specs_handler: public specs_setter<typename Context::char_type> {
-  arg_ref<Char> width_ref;
-  arg_ref<Char> precision_ref;
+  typedef arg_ref<Char, string_value<Char>> arg_ref_type;
+  arg_ref_type width_ref;
+  arg_ref_type precision_ref;
 };
 
 // Format spec handler that saves references to arguments representing dynamic
 // width and precision to be resolved at formatting time.
-template <typename ParseContext>
-class dynamic_specs_handler :
-    public specs_setter<typename ParseContext::char_type> {
+template <typename Specs, typename ParseContext, typename ArgRefCreator>
+class dynamic_specs_handler
+    : public specs_setter<typename ParseContext::char_type> {
  public:
   typedef typename ParseContext::char_type char_type;
 
-  FMT_CONSTEXPR dynamic_specs_handler(
-      dynamic_format_specs<char_type> &specs, ParseContext &ctx)
-    : specs_setter<char_type>(specs), specs_(specs), context_(ctx) {}
+  FMT_CONSTEXPR dynamic_specs_handler(Specs &specs, ParseContext &ctx,
+                                      ArgRefCreator &creator)
+      : specs_setter<char_type>(specs),
+        specs_(specs),
+        context_(ctx),
+        arg_ref_creator_(creator) {}
 
   FMT_CONSTEXPR dynamic_specs_handler(const dynamic_specs_handler &other)
-    : specs_setter<char_type>(other),
-      specs_(other.specs_), context_(other.context_) {}
+      : specs_setter<char_type>(other),
+        specs_(other.specs_),
+        context_(other.context_),
+        arg_ref_creator_(other.arg_ref_creator_) {}
 
   template <typename Id>
   FMT_CONSTEXPR void on_dynamic_width(Id arg_id) {
-    specs_.width_ref = make_arg_ref(arg_id);
+    context_.check_arg_id(arg_id);
+    specs_.width_ref = arg_ref_creator_.create(arg_id);
   }
 
   template <typename Id>
   FMT_CONSTEXPR void on_dynamic_precision(Id arg_id) {
-    specs_.precision_ref = make_arg_ref(arg_id);
+    context_.check_arg_id(arg_id);
+    specs_.precision_ref = arg_ref_creator_.create(arg_id);
   }
 
   FMT_CONSTEXPR void on_error(const char *message) {
     context_.on_error(message);
   }
 
  private:
-  typedef arg_ref<char_type> arg_ref_type;
+  Specs &specs_;
+  ParseContext &context_;
+  ArgRefCreator &arg_ref_creator_;
+};
 
-  template <typename Id>
-  FMT_CONSTEXPR arg_ref_type make_arg_ref(Id arg_id) {
-    context_.check_arg_id(arg_id);
-    return arg_ref_type(arg_id);
+template <typename ParseContext>
+class dynamic_arg_ref_creator {
+  typedef typename ParseContext::char_type char_type;
+
+ public:
+  typedef arg_ref<char_type, string_value<char_type>> arg_ref_type;
+
+  FMT_CONSTEXPR dynamic_arg_ref_creator(ParseContext &ctx) : context_(ctx) {}
+
+  FMT_CONSTEXPR arg_ref_type create(unsigned id) { return arg_ref_type(id); }
+
+  FMT_CONSTEXPR arg_ref_type create(basic_string_view<char_type> id) {
+    return arg_ref_type(string_value<char_type>{id.data(), id.size()});
   }
 
-  FMT_CONSTEXPR arg_ref_type make_arg_ref(auto_id) {
+  FMT_CONSTEXPR arg_ref_type create(auto_id) {
     return arg_ref_type(context_.next_arg_id());
```

Comment:
I think `context_.next_arg_id()` should be called in `dynamic_specs_handler` and the ID passed to `create`. Then `dynamic_arg_ref_creator` won't depend on parse context.